src/tox/config/main.py

```diff
@@ -159,18 +160,21 @@ def get_env(
         :param loaders: loaders to use for this configuration (only used for creation)
         :return: the tox environments config
         """
-        section, base_test, base_pkg = self._src.get_tox_env_section(item)
-        return self.get_section_config(
-            section,
-            base=base_pkg if package else base_test,
-            of_type=EnvConfigSet,
-            for_env=item,
-            loaders=loaders,
-        )
+        if item not in self._key_to_env_conf:
+            section, base_test, base_pkg = self._src.get_tox_env_section(item)
+            self._key_to_env_conf[item] = self.get_section_config(
+                section,
+                base=base_pkg if package else base_test,
+                of_type=EnvConfigSet,
+                for_env=item,
+                loaders=loaders,
+            )
+        return self._key_to_env_conf[item]
 
     def clear_env(self, name: str) -> None:
         section, _, __ = self._src.get_tox_env_section(name)
         self._key_to_conf_set = {k: v for k, v in self._key_to_conf_set.items() if k[0] == section.key and k[1] == name}
```

Comment:
@gaborbernat this line looks wrong to me - surely the `_key_to_conf_set` should be set to those items where the key is not equal to the name being cleared? I think that it probably isn't easy to notice the issue because mostly the deleted conf would be recreated when next requested, and because (I think) the env that is being cleared will not clash with the replacement because it would only ever be a run env being deleted and replaced with a packaging env, so they wouldn't match keys to `_key_to_conf_set`.